### DISCUSSION TITLE
Any way to stop traversal?

### DISCUSSION THREAD

**Discussion by mattvague** [josdejong/mathjs#2588] [Q&A]
After looking at [the code](https://github.com/josdejong/mathjs/blob/develop/src/expression/node/Node.js#L117-L131) I'm pretty sure the answer to this is no, but just wanted to make sure: there's currently no way to stop traversal of a tree once `node.traverse(...)` has been called, correct? 

My current use case is that I want to do a tree search for functions of a particular variable, but stopping on the first match. I have something like:

```js
let match = false

node.traverse((child) => {
  if (isFunctionNode(child) && child.args.some(containsVariable(myVar)) {
    match = child
  }
})

return match
```

But this is not ideal as it will continue to traverse the tree in it's entirety wasting cycles and possibly giving me the wrong result. What would be nicer is if I could do something like this to stop early:

```js
let match = false

node.traverse((child) => {
  if (isFunctionNode(child) && child.args.some(containsVariable(myVar)) {
    match = child
    return false
  }
})

return match
```

@josdejong If I'm correct that this is currently not possible, would you accept a PR for it?

**Comment by josdejong** [comment]
Thanks, that is an interesting point. 

Currently, `traverse` is basically a recursive version of `forEach`, and in both cases there is no way to stop under specific conditions. In the same way, `transform` is the recursive version of `map`. But `transform` is a bit different: this allows replacing a node with something new, in which case the iteration is stopped. So a workaround for now could be to "misuse" `transform` to stop early, by returning something like `null` or `false` as soon as your match is found.

However, maybe `traverse` is not the best solution anyway: what you're doing is finding a match. So maybe we should consider implementing new recursive methods like `find` and `findAll`. Shall we discuss that in detail in #2589?

EDIT: I think we already have a `findAll`, it is called `filter` 😁
